messaging/src/main/java/org/axonframework/eventhandling/pooled/WorkPackage.java

```diff
@@ -300,38 +303,51 @@ private void processEvents() throws Exception {
         if (!eventBatch.isEmpty()) {
             logger.debug("Work Package [{}]-[{}] is processing a batch of {} events.",
                          segment.getSegmentId(), name, eventBatch.size());
-            UnitOfWork<TrackedEventMessage<?>> unitOfWork = new BatchingUnitOfWork<>(eventBatch);
-            unitOfWork.attachTransaction(transactionManager);
-            unitOfWork.resources().put(segmentIdResourceKey, segment.getSegmentId());
-            unitOfWork.resources().put(lastTokenResourceKey, lastConsumedToken);
-            unitOfWork.onPrepareCommit(u -> storeToken(lastConsumedToken));
-            unitOfWork.afterCommit(
-                    u -> segmentStatusUpdater.accept(status -> status.advancedTo(lastConsumedToken))
-            );
-            batchProcessor.processBatch(eventBatch, unitOfWork, Collections.singleton(segment));
+            try {
+                processingEvents.set(true);
+                UnitOfWork<TrackedEventMessage<?>> unitOfWork = new BatchingUnitOfWork<>(eventBatch);
+                unitOfWork.attachTransaction(transactionManager);
+                unitOfWork.resources().put(segmentIdResourceKey, segment.getSegmentId());
+                unitOfWork.resources().put(lastTokenResourceKey, lastConsumedToken);
+                unitOfWork.onPrepareCommit(u -> storeToken(lastConsumedToken));
+                unitOfWork.afterCommit(
+                        u -> segmentStatusUpdater.accept(status -> status.advancedTo(lastConsumedToken))
+                );
+                batchProcessor.processBatch(eventBatch, unitOfWork, Collections.singleton(segment));
+            } finally {
+                processingEvents.set(false);
+            }
         } else {
             segmentStatusUpdater.accept(status -> status.advancedTo(lastConsumedToken));
-            if (lastClaimExtension < clock.instant().toEpochMilli() - claimExtensionThreshold) {
-                if (lastStoredToken != lastConsumedToken) {
-                    transactionManager.executeInTransaction(() -> storeToken(lastConsumedToken));
-                } else {
-                    transactionManager.executeInTransaction(this::extendClaim);
-                }
+            if (claimExtensionThresholdMet() && lastStoredToken != lastConsumedToken) {
```

Comment:
This does change behavior even if they don't set the new flag. The token may be extended fewer times when they are few events, that's still an improvement, so it should be fine.

Reply:
I believe it stays the same. The outcome, at least.

The `extendClaimIfThresholdIsMet` performs the `claimExtensionThresholdMet` itself too.

So, whenever `claimExtensionThresholdMet` renders true and the `lastConsumedToken` is different, we store the token (same as before).

Whenever the `lastConsumedToken` is identical, we check the `claimExtensionThresholdMet` inside the `extendClaimIfThresholdIsMet` operation.
This essentially flips the order of validation, as previously we first checked the timeout and then whether the tokens were identical.

If the `lastConsumedToken` returns `true`, but `claimExtensionThresholdMet` returns false, that means we'll move to the `else`, which will make the `claimExtensionThresholdMet` again.
*This* is the biggest change, I think, as we're potentially duplicating the `claimExtensionThresholdMet` check.

Reply:
Indeed, the check for `lastClaimExtension < clock.instant().toEpochMilli() - claimExtensionThreshold` was already there.